### ISSUE TITLE
[Attention][torch.compile] All false attention mask causes output to be NaN after attention fusion

### ISSUE THREAD

**Issue by YangChenyuan** [pytorch/pytorch#100318]
### 🐛 Describe the bug

All false attention mask causes output to be NaN after attention fusion. This incorrect and unexpected output can cause security risks in real applications using such compiled LLMs.

The following is the minimized code for the vulnerability. The model contains attention computation which will be optimized by `fuse_attention` in `torch.compile/_inductor`.

In the code example, we provide an all-false attention mask `attn_mask`. The model without compilation returns the correct value. By contrast, the model (`jit_func`) after the optimization of `fuse_attention` will return NaN as output.

Notably, this bug is only revealed when the `attn_mask` is all-false. If we replace the `attn_mask` with `torch.ones(batch_size, num_heads, sequence_length, sequence_length).bool()`, the results will be consistent.

```py
import math
import torch

torch.manual_seed(420)

input_dim = 4
sequence_length = 3
batch_size = 2
num_heads = 2
hidden_dim = 1
dropout_p = 1

inputs = torch.randn(batch_size, sequence_length, input_dim)
attn_mask = torch.zeros(batch_size, num_heads, sequence_length, sequence_length).bool()

class Attention(torch.nn.Module):

    def __init__(self, input_dim, hidden_dim, num_heads, dropout_p):
        super().__init__()
        self.query = torch.nn.Linear(input_dim, hidden_dim * num_heads, bias=False)
        self.key = torch.nn.Linear(input_dim, hidden_dim * num_heads, bias=False)
        self.value = torch.nn.Linear(input_dim, hidden_dim * num_heads, bias=False)
        self.dropout = torch.nn.Dropout(dropout_p)
        self.num_heads = num_heads
        self.hidden_dim = hidden_dim

    def forward(self, inputs, attn_mask=None):
        q = self.query(inputs).view(inputs.size(0), self.num_heads, -1, self.hidden_dim)
        k = self.key(inputs).view(inputs.size(0), self.num_heads, -1, self.hidden_dim)
        v = self.value(inputs).view(inputs.size(0), self.num_heads, -1, self.hidden_dim)
        scores = q @ k.transpose(-2, -1) / math.sqrt(self.hidden_dim)
        scores = scores + attn_mask
        attn_weights = torch.softmax(scores, dim=-1)
        output = attn_weights @ v
        return output

func = Attention(input_dim, hidden_dim, num_heads, dropout_p).to('cpu')

res1 = func(inputs, attn_mask)
print(res1)
# tensor([[[[ 0.0261],
#           [ 0.0116],
#           [ 0.0387]],
# 
#          [[-0.0424],
#           [-0.0398],
#           [-0.0697]]],
# 
# 
#         [[[-0.1785],
#           [-0.1944],
#           [-0.1605]],
# 
#          [[-0.5129],
#           [-0.5133],
#           [-0.5141]]]], grad_fn=<UnsafeViewBackward0>)  

jit_func = torch.compile(func)
res2 = jit_func(inputs, attn_mask)
print(res2)
# tensor([[[[nan],
#           [nan],
#           [nan]],
# 
#          [[nan],
#           [nan],
#           [nan]]],
# 
# 
#         [[[nan],
#           [nan],
#           [nan]],
# 
#          [[nan],
#           [nan],
#           [nan]]]], grad_fn=<CompiledFunctionBackward>)
```

The attacker can use this vulnerability to manipulate the model output. Attention is widely used in large language models.
By providing a special attention mask (all False) to the model, attackers can make the output to be all NaN.

I believe this is a security vulnerability and have reported this issue to https://github.com/pytorch/pytorch/security/advisories/new

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230428+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230428+cu118
[pip3] torchaudio==2.1.0.dev20230428+cu118
[pip3] torchvision==0.16.0.dev20230428+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230428+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230428+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230428+cu118          pypi_0    pypi
```

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

**Comment by YangChenyuan** [comment]
Is there any update for this bug?
